tests/helpers/test_trigger.py

```diff
@@ -780,3 +790,229 @@ async def good_subscriber(new_triggers: set[str]):
+    hass: HomeAssistant,
+) -> None:
+    """Test async_track_target_selector_state_change_event with multiple targets."""
+    calls = []
+
+    @callback
+    def state_change_callback(event):
+        """Handle state change events."""
+        calls.append(event)
+
+    async def set_state(entity_id, state):
+        """Set the state of an entity."""
+        hass.states.async_set(entity_id, state)
+        await hass.async_block_till_done()
+
+    def assert_entity_calls_and_reset(entity_id: str) -> None:
+        assert len(calls) == 1
+        assert calls[0].data["entity_id"] == entity_id
+        calls.clear()
+
+    config_entry = MockConfigEntry(domain="test")
+    config_entry.add_to_hass(hass)
+
+    device_reg = dr.async_get(hass)
+    device_entry = device_reg.async_get_or_create(
+        config_entry_id=config_entry.entry_id,
+        identifiers={("test", "device_1")},
+    )
+
+    untargeted_device_entry = device_reg.async_get_or_create(
+        config_entry_id=config_entry.entry_id,
+        identifiers={("test", "area_device")},
+    )
+
+    entity_reg = er.async_get(hass)
+    device_entity = entity_reg.async_get_or_create(
+        domain="light",
+        platform="test",
+        unique_id="device_light",
+        device_id=device_entry.id,
+    ).entity_id
+
+    untargeted_device_entity = entity_reg.async_get_or_create(
+        domain="light",
+        platform="test",
+        unique_id="area_device_light",
+        device_id=untargeted_device_entry.id,
+    ).entity_id
+
+    untargeted_entity = entity_reg.async_get_or_create(
+        domain="light",
+        platform="test",
+        unique_id="untargeted_light",
+    ).entity_id
+
+    targeted_entity = "light.test_light"
+
+    for entity_id in (targeted_entity, device_entity, untargeted_entity):
+        hass.states.async_set(entity_id, STATE_OFF)
+    await hass.async_block_till_done()
+
+    label = lr.async_get(hass).async_create("Test Label").name
+    area = ar.async_get(hass).async_create("Test Area").id
+    floor = fr.async_get(hass).async_create("Test Floor").floor_id
+
+    selector_config = {
+        ATTR_ENTITY_ID: targeted_entity,
+        ATTR_DEVICE_ID: device_entry.id,
+        ATTR_AREA_ID: area,
+        ATTR_FLOOR_ID: floor,
+        ATTR_LABEL_ID: label,
+    }
+    unsub = async_track_target_selector_state_change_event(
+        hass, selector_config, state_change_callback
+    )
+
+    # Test directly targeted entity and device
+    await set_state(targeted_entity, STATE_ON)
+    await set_state(device_entity, STATE_ON)
+
+    assert len(calls) == 2
+    assert calls[0].data["entity_id"] == targeted_entity
+    assert calls[0].data["old_state"].state == STATE_OFF
+    assert calls[0].data["new_state"].state == STATE_ON
+    assert calls[1].data["entity_id"] == device_entity
+    assert calls[1].data["old_state"].state == STATE_OFF
+    assert calls[1].data["new_state"].state == STATE_ON
+    calls.clear()
```

Comment:
I'd suggest to adjust the test a bit to always fire state changes for all the three entities, to make sure the correct subscriptions are maintained.
Maybe adjust set_state to set the state of all entities, and also adjust assert_entity_calls_and_reset to accept a list of entity_ids?